### PR TITLE
Feature/QueryGroup Add Create API in plugin

### PR DESCRIPTION
**Pull request by ruai0511**
### Description
This change introduces the Create QueryGroup API which we will use to enforce node level resiliency as part of this RFC: https://github.com/opensearch-project/OpenSearch/issues/12342.

The Create QueryGroup API schema is:
```
curl -XPUT "localhost:9200/_query_group/" -H 'Content-Type: application/json' -d ' 
{
     "name": "analytics",
     "jvm":0.4,
     "mode": "monitor"
}'
```

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by andrross** [comment]
A couple high level comments:

**Comment by andrross** [comment]
A couple high level comments:

- This PR is creating a new plugin in the core repo, contrary to the (recently introduced) [convention](https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md#creating-new-plugins). I don't see it mentioned in the RFC, but please explain (or link to an explanation) of why this plugin should be in the core repo as opposed to a separate repo.
- Please look at #14456 as this PR appears to be introducing substantial changes to the REST API. I would love to see these new APIs defined and documented in https://github.com/opensearch-project/opensearch-api-specification in a draft PR before reviewing the implementation.

**Comment by ruai0511** [comment]
> A couple high level comments:
> 
> * This PR is creating a new plugin in the core repo, contrary to the (recently introduced) [convention](https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md#creating-new-plugins). I don't see it mentioned in the RFC, but please explain (or link to an explanation) of why this plugin should be in the core repo as opposed to a separate repo.
> * Please look at [[PROPOSAL] Highlight REST API Changes in PRs  #14456](https://github.com/opensearch-project/OpenSearch/issues/14456) as this PR appears to be introducing substantial changes to the REST API. I would love to see these new APIs defined and documented in https://github.com/opensearch-project/opensearch-api-specification in a draft PR before reviewing the implementation.

Thanks for the info! I'll look into it

**Comment by kaushalmahi12** [comment]
Thanks @andrross for going through the PR and provide useful insights.  I have updated the RFC for the reasoning behind moving this into a plugin. We had also discussed this with @msfroh.

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/CreateQueryGroupResponse.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field("name", queryGroup.getName());
+        builder.field("mode", queryGroup.getMode().getName());
+        builder.field("updatedAt", queryGroup.getUpdatedAtInMillis());
+        builder.mapContents(queryGroup.getResourceLimits());
+        builder.endObject();
```
[the remark this answers is not in the document]
Comment (kaushalmahi12) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/CreateQueryGroupResponse.java:63]: +1

File: plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/QueryGroupTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            settings,
+            clusterSettings
+        );
+        return List.of(queryGroupPersistenceService, clusterState);
+    }
+
+    public static void compareResourceLimits(Map<String, Object> resourceLimitMapOne, Map<String, Object> resourceLimitMapTwo) {
+        for (String resourceName : resourceLimitMapOne.keySet()) {
+            assertTrue(resourceLimitMapTwo.containsKey(resourceName));
+            assertEquals(resourceLimitMapOne.get(resourceName), resourceLimitMapTwo.get(resourceName));
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/QueryGroupTestUtils.java:97]: We can replace this with

```suggestion
        assertTrue(resourceLimitMapOne.keySet().containsAll(resourceLimitMapTwo.keySet()));
        assertTrue(resourceLimitMapOne.values().containsAll(resourceLimitMapTwo.values()));
```

File: plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/QueryGroupTestUtils.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            String groupOneName = groupOne.getName();
+            List<QueryGroup> groupTwoList = listTwo.stream().filter(sb -> sb.getName().equals(groupOneName)).collect(Collectors.toList());
+            assertEquals(1, groupTwoList.size());
+            QueryGroup groupTwo = groupTwoList.get(0);
+            assertEquals(groupOne.getName(), groupTwo.getName());
+            assertEquals(groupOne.get_id(), groupTwo.get_id());
+            compareResourceLimits(groupOne.getResourceLimits(), groupTwo.getResourceLimits());
+            assertEquals(groupOne.getMode(), groupTwo.getMode());
+            assertEquals(groupOne.getUpdatedAtInMillis(), groupTwo.getUpdatedAtInMillis());
+        }
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/test/java/org/opensearch/plugin/wlm/QueryGroupTestUtils.java:113]: How about sorting these two lists and then doing the index wise comparison I think that would be more efficient ?
If the query group already implements `equalsTo` method then we should leverage that.

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        double existingUsage = 0;
+        for (QueryGroup group : groupsList) {
+            if (!group.getName().equals(groupName)) {
+                existingUsage += getResourceLimitValue(resourceName, group.getResourceLimits());
+            }
+        }
+        return existingUsage;
+    }
+
+    List<QueryGroup> getFromClusterStateMetadata(String name, ClusterState currentState) {
```
Comment (kaushalmahi12) [inline plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java:210]: I don't see the use of this method ?

File: plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/CreateQueryGroupResponse.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import java.io.IOException;
+
+/**
+ * Response for the create API for QueryGroup
+ *
+ * @opensearch.internal
+ */
+public class CreateQueryGroupResponse extends ActionResponse implements ToXContent, ToXContentObject {
+    private final QueryGroup queryGroup;
+    private RestStatus restStatus;
```
[the remark this answers is not in the document]
Comment (ruai0511) [reply plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/CreateQueryGroupResponse.java:29]: This one is NotSerializableExceptionWrapper. But I also see other classes that don't store reststatus and only store data.

**Comment by ruai0511** [comment]
Move this to https://github.com/opensearch-project/OpenSearch/pull/14680

**Comment by ruai0511** [comment]
Move this to https://github.com/opensearch-project/OpenSearch/pull/14680 and close this PR

**Pull request closed without merging**